coderd/database/dbfake/dbfake.go

```diff
@@ -0,0 +1,124 @@
+	t.Helper()
+
+	// This intentionally fulfills the minimum requirements of the schema.
+	// Tests can provide a custom template ID if necessary.
+	if seed.TemplateID == uuid.Nil {
+		template := dbgen.Template(t, db, database.Template{
+			OrganizationID: seed.OrganizationID,
+			CreatedBy:      seed.OwnerID,
+		})
+		seed.TemplateID = template.ID
+		seed.OwnerID = template.CreatedBy
+		seed.OrganizationID = template.OrganizationID
+	}
+	return dbgen.Workspace(t, db, seed)
+}
+
+// CreateWorkspaceWithAgent is a helper that generates a workspace with a single resource
+// that has an agent attached to it. The agent token is returned.
+func CreateWorkspaceWithAgent(t testing.TB, db database.Store, seed database.Workspace) (database.Workspace, string) {
+	t.Helper()
+	authToken := uuid.NewString()
+	ws := CreateWorkspace(t, db, seed)
+	CreateWorkspaceBuild(t, db, ws, database.WorkspaceBuild{}, &proto.Resource{
+		Name: "example",
+		Type: "aws_instance",
+		Agents: []*proto.Agent{{
+			Id: uuid.NewString(),
+			Auth: &proto.Agent_Token{
+				Token: authToken,
+			},
+		}},
+	})
+	return ws, authToken
+}
+
+// CreateWorkspaceBuild inserts a build and a successful job into the database.
+func CreateWorkspaceBuild(t testing.TB, db database.Store, ws database.Workspace, seed database.WorkspaceBuild, resources ...*sdkproto.Resource) database.WorkspaceBuild {
+	t.Helper()
+	jobID := uuid.New()
+	seed.JobID = jobID
+	seed.WorkspaceID = ws.ID
+	// This intentionally fulfills the minimum requirements of the schema.
+	// Tests can provide a custom version ID if necessary.
+	if seed.TemplateVersionID == uuid.Nil {
+		jobID := uuid.New()
+		templateVersion := dbgen.TemplateVersion(t, db, database.TemplateVersion{
+			JobID:          jobID,
+			OrganizationID: ws.OrganizationID,
+			CreatedBy:      ws.OwnerID,
+			TemplateID: uuid.NullUUID{
+				UUID:  ws.TemplateID,
+				Valid: true,
+			},
+		})
+		payload, _ := json.Marshal(provisionerdserver.TemplateVersionImportJob{
+			TemplateVersionID: templateVersion.ID,
+		})
+		dbgen.ProvisionerJob(t, db, nil, database.ProvisionerJob{
+			ID:             jobID,
+			OrganizationID: ws.OrganizationID,
+			Input:          payload,
+			Type:           database.ProvisionerJobTypeTemplateVersionImport,
+			CompletedAt: sql.NullTime{
+				Time:  dbtime.Now(),
+				Valid: true,
+			},
+		})
+		seed.TemplateVersionID = templateVersion.ID
+	}
+	build := dbgen.WorkspaceBuild(t, db, seed)
+
+	payload, err := json.Marshal(provisionerdserver.WorkspaceProvisionJob{
+		WorkspaceBuildID: build.ID,
+	})
+	require.NoError(t, err)
+	job := dbgen.ProvisionerJob(t, db, nil, database.ProvisionerJob{
+		ID:             jobID,
+		Input:          payload,
+		OrganizationID: ws.OrganizationID,
+		CompletedAt: sql.NullTime{
+			Time:  dbtime.Now(),
+			Valid: true,
+		},
+	})
+	CreateProvisionerJobResources(t, db, job.ID, seed.Transition, resources...)
+	return build
+}
```

Comment:
This feels like we will have `CreateWorkspaceWithTwoAgents`, `CreateWorkspaceWithParameters`, etc etc.

Instead, can we add modifiers to `CreateWorkspaceBuild`? Like:


```golang
dbfake.CreateWorkspaceBuild(t, db, ws, database.WorkspaceBuild{}, dbfake.AgentResource(/*params*/), dbfake.AgentResource(/*params*/), dbfake.MetaData(/*params*/))
```

So `CreateWorkspaceWithAgent` is just the WorkspaceBuild with some resource function to create the "default" agent?

Reply:
I'm not sure it's bad to have root helpers. The return arguments are different, so we can't do that pattern exactly.

Reply:
If it is super common, then sure. For that one, I figured you could just pass a random string in.

```golang
const authToken = "something"
dbfake.CreateWorkspaceBuild(t, db, ws, database.WorkspaceBuild{}, dbfake.AgentResource(authToken))
```

Reply:
It looks like it is common though, so fine to keep. Just thinking of ways to generalize the api